FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/mqtt_operations.c

```diff
@@ -0,0 +1,1069 @@
+
+    if( democonfigMQTT_BROKER_PORT == 443 )
+    {
+        /* Pass the ALPN protocol name depending on the port being used.
+         * Please see more details about the ALPN protocol for AWS IoT MQTT endpoint
+         * in the link below.
+         * https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/
+         */
+        xTlsCredentials.pAlpnProtos = pcAlpn;
+    }
+
+    /* Initialize reconnect attempts and interval */
+    BackoffAlgorithm_InitializeParams( &xReconnectParams,
+                                       mqttopCONNECTION_RETRY_BACKOFF_BASE_MS,
+                                       mqttopCONNECTION_RETRY_MAX_BACKOFF_DELAY_MS,
+                                       mqttopCONNECTION_RETRY_MAX_ATTEMPTS );
+
+    do
+    {
+        /* Establish a TLS session with the MQTT broker. This example connects
+         * to the MQTT broker as specified in democonfigMQTT_BROKER_ENDPOINT and democonfigMQTT_BROKER_PORT
+         * at the demo config header. */
+        LogDebug( ( "Establishing a TLS session to %.*s:%d.",
+                    democonfigMQTT_BROKER_ENDPOINT_LENGTH,
+                    democonfigMQTT_BROKER_ENDPOINT,
+                    democonfigMQTT_BROKER_PORT ) );
+
+        xTlsStatus = TLS_FreeRTOS_Connect( pxNetworkContext,
+                                           democonfigMQTT_BROKER_ENDPOINT,
+                                           democonfigMQTT_BROKER_PORT,
+                                           &xTlsCredentials,
+                                           mqttopTRANSPORT_SEND_RECV_TIMEOUT_MS, mqttopTRANSPORT_SEND_RECV_TIMEOUT_MS );
+
+        if( xTlsStatus == TLS_TRANSPORT_SUCCESS )
+        {
+            /* Connection successful. */
+            xReturnStatus = true;
+        }
+        else
+        {
+            /* Generate a random number and get back-off value (in milliseconds) for the next connection retry. */
+            xBackoffAlgStatus = BackoffAlgorithm_GetNextBackoff( &xReconnectParams, prvGenerateRandomNumber(), &usNextRetryBackOff );
+
+            if( xBackoffAlgStatus == BackoffAlgorithmRetriesExhausted )
+            {
+                LogError( ( "Connection to the broker failed, all attempts exhausted." ) );
+            }
+            else if( xBackoffAlgStatus == BackoffAlgorithmSuccess )
+            {
+                LogWarn( ( "Connection to the broker failed. Retrying connection "
+                           "after %hu ms backoff.",
+                           ( unsigned short ) usNextRetryBackOff ) );
+                vTaskDelay( pdMS_TO_TICKS( usNextRetryBackOff ) );
+            }
+        }
+    } while( ( xTlsStatus != TLS_TRANSPORT_SUCCESS ) && ( xBackoffAlgStatus == BackoffAlgorithmSuccess ) );
+
+    return xReturnStatus;
+}
+/*-----------------------------------------------------------*/
+
+static bool prvGetNextFreeIndexForOutgoingPublishes( uint8_t * pucIndex )
```

Comment:
It's not great practice to use a global in this way. It would be better to include pxOutgoingPublishPackets as a parameter to this function and other similar functions.

Reply:
While it might be better practice to modify this, all of the demos use the same design pattern with the global in both FreeRTOS and the CSDK - [see here](https://github.com/FreeRTOS/FreeRTOS/blob/main/FreeRTOS-Plus/Demo/AWS/Mqtt_Demo_Helpers/mqtt_demo_helpers.c#L245-L250).

If we want to change this, we should leave it for another task and change all demos consistently.

Reply:
Perhaps this should be part of the library if we have so many copies of the same file :).

Reply:
The other AWS demos share the "mqtt_demo_helpers" files. However, those use "using_mbedtls.h" rather than "using_mbedtls_pkcs11.h" (and has other incompatibilities with pkcs11), so it wasn't compatible with the demo.